File: pact_test.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"testing"

	"github.com/google/martian/log"
	"github.com/pact-foundation/pact-go/dsl"
	"github.com/pact-foundation/pact-go/types"
	"github.com/stretchr/testify/assert"
)

func GetChatData() ProducerResponseWrapper {
	return ProducerResponseWrapper{
		Data: ProducerResponseData{
			Type: "ChatTemplate",
			Attributes: ProducerResponseAttributes{
				ProducerChatTemplates: []ProducerChatTemplate{
					{
						Id:       "1",
						Consumer: "fullserve",
						Title:    "title1",
						Category: "category1",
						Message:  "message1",
						Queues:   []string{"queue1", "queue2"},
					},
					{
						Id:       "2",
						Consumer: "fullserve",
						Title:    "title2",
						Category: "category2",
						Message:  "message2",
						Queues:   []string{"queue3", "queue4"},
					},
					{
						Id:       "3",
						Consumer: "fullserve",
						Title:    "title3",
						Category: "category3",
						Message:  "message3",
						Queues:   []string{"queue5"},
					},
					{
						Id:       "4",
						Consumer: "fullserve",
						Title:    "title4",
						Category: "category4",
						Message:  "message4",
						Queues:   []string{"queue6", "queue7", "queue8"},
					},
				},
			},
		},
	}
}

func hitChatDataAPI() (err error) {
	url := fmt.Sprintf("http://localhost:%d/actions/bcd/chat-template", pact.Server.Port)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Content-Type", "application/json")

	http.DefaultClient.Do(req)

	return
}

var pact dsl.Pact
var dir, _ = os.Getwd()
var pactDir = fmt.Sprintf("%s/pacts", dir)
var logDir = fmt.Sprintf("%s/logs", dir)

func TestMain(m *testing.M) {
	// Setup Pact and related test stuff
	pact = dsl.Pact{
		Consumer:                 "toolbar-chat-api",
		Provider:                 "unv-bcd-chat-pdr",
		DisableToolValidityCheck: true,
		PactDir:                  pactDir,
		LogDir:                   logDir,
	}

	// Proactively start service to get access to the port
	pact.Setup(true)

	// Run all the tests
	code := m.Run()

	// Shutdown the Mock Service and write pact files to disk
	err := pact.WritePact()
	if err != nil {
		log.Infof("Failed to write your contract")
		return
	}

	pact.Teardown()
	os.Exit(code)
}

func TestConsumerPact(t *testing.T) {
	chatData := GetChatData()
	pact.AddInteraction().Given("Chat data of fullserve consumer exists").
		UponReceiving("A request to get chat data").
		WithRequest(dsl.Request{
			Method: "GET",
			Path:   dsl.Term("/actions/bcd/chat-template", "/actions/bcd/chat-template"),
			Headers: dsl.MapMatcher{
				"Content-Type": dsl.Term("application/json", `application\/json`),
			},
		}).
		WillRespondWith(dsl.Response{
			Status: 200,
			Headers: dsl.MapMatcher{
				"Content-Type": dsl.Term("application/json", `application\/json`),
			},
			Body: chatData,
		})
	err := pact.Verify(hitChatDataAPI)
	if err != nil {
		t.Fatalf("Error on Verify: %v", err)
	}
	assert.NoError(t, err)

	// Store contract remotely
	publisher := dsl.Publisher{}
	err = publisher.Publish(types.PublishRequest{
		PactURLs:        []string{"pacts/toolbar-chat-api-unv-bcd-chat-pdr.json"},
		PactBroker:      "https://harsh.pactflow.io/", //link to your remote Contract broker
		BrokerToken:     "2_KfMXbOMRXKAd30PwopTg",     //your PactFlow token
		ConsumerVersion: "1.0.0",
		Tags:            []string{"1.0.0", "latest"},
	})
	if err != nil {
		t.Fatal(err)
	}
	assert.NoError(t, err)
}

File: pact_test.py
import atexit
import unittest
from pact import Consumer, Provider, Like, EachLike
import requests
import os
import warnings

# Suppress Deprecation Warnings (Temporary Measure)
warnings.filterwarnings("ignore", category=DeprecationWarning)

os.environ['NO_PROXY'] = 'localhost,127.0.0.1'

# Define Pact between the Consumer and Provider
pact = Consumer('toolbar-chat-api').has_pact_with(
    Provider('unv-bcd-chat-pdr'),
    pact_dir='./pacts',
    port=1235
)
pact.start_service()
atexit.register(pact.stop_service)  # Ensure service is stopped at exit


class GetChatDataConsumerTest(unittest.TestCase):

    def setUp(self):
        """Set up before each test case."""
        self.pact = pact

    def tearDown(self):
        """Clean up after each test case."""
        pass  # The mock service is already stopped by atexit

    def define_interaction(self, description, consumer_type, expected_response, status_code=200, auth_required=True, token=None):
        """Helper method to define Pact interactions."""
        headers = {}
        if auth_required and token:
            headers = {"Authorization": f"Bearer {token}"}

        # Define the expected interaction with the Provider
        (self.pact
         .given(f'Chat data of {consumer_type} consumer exists')
         .upon_receiving(description)
         .with_request(
             'GET',
             '/bcd-dev/actions/bcd/chat-template',
             headers=headers,
             query={'consumer': consumer_type}
         )
         .will_respond_with(status_code, body=expected_response))

    def test_get_chat_data_fullserve(self):
        """Test case for consumer 'fullserve'."""
        consumer_type = 'fullserve'
        expected_interaction = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": EachLike(
                        {
                            "id": Like("2"),
                            "consumer": "fullserve",
                            "title": Like("title2"),
                            "category": Like("category2"),
                            "message": Like("message2"),
                            "queues": EachLike("queue3")
                        }
                    )
                }
            }
        }

        # Define the expected concrete response for assertions
        expected_response = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": [
                        {
                            "id": "2",
                            "consumer": "fullserve",
                            "title": "title2",
                            "category": "category2",
                            "message": "message2",
                            "queues": ["queue3"]
                        }
                    ]
                }
            }
        }

        # Manually provide the token
        token = 'your_dynamic_token_here'

        # Define interaction
        self.define_interaction(
            description='A request to get chat data for fullserve',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            token=token
        )

        # Run the test: Make the actual request
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}', headers={"Authorization": f"Bearer {token}"})

        # Assert that the result matches the expectation
        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), expected_response)

    def test_get_chat_data_veripark(self):
        """Test case for consumer 'veripark'."""
        consumer_type = 'veripark'
        expected_interaction = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": EachLike(
                        {
                            "id": Like("2"),
                            "consumer": "veripark",
                            "title": Like("title2"),
                            "category": Like("category2"),
                            "message": Like("message2"),
                            "queues": EachLike("queue3")
                        }
                    )
                }
            }
        }

        # Define the expected concrete response for assertions
        expected_response = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": [
                        {
                            "id": "2",
                            "consumer": "veripark",
                            "title": "title2",
                            "category": "category2",
                            "message": "message2",
                            "queues": ["queue3"]
                        }
                    ]
                }
            }
        }

        # Manually provide the token
        token = 'your_dynamic_token_here'

        # Define interaction
        self.define_interaction(
            description='A request to get chat data for veripark',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            token=token
        )

        # Run the test: Make the actual request
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}', headers={"Authorization": f"Bearer {token}"})

        # Assert that the result matches the expectation
        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), expected_response)

    def test_get_chat_data_salesforce(self):
        """Test case for consumer 'salesforce'."""
        consumer_type = 'salesforce'
        expected_interaction = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": EachLike(
                        {
                            "id": Like("2"),
                            "consumer": "salesforce",
                            "title": Like("title2"),
                            "category": Like("category2"),
                            "message": Like("message2"),
                            "queues": EachLike("queue3")
                        }
                    )
                }
            }
        }

        # Define the expected concrete response for assertions
        expected_response = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": [
                        {
                            "id": "2",
                            "consumer": "salesforce",
                            "title": "title2",
                            "category": "category2",
                            "message": "message2",
                            "queues": ["queue3"]
                        }
                    ]
                }
            }
        }

        # Manually provide the token
        token = 'your_dynamic_token_here'

        # Define interaction
        self.define_interaction(
            description='A request to get chat data for salesforce',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            token=token
        )

        # Run the test: Make the actual request
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}', headers={"Authorization": f"Bearer {token}"})

        # Assert that the result matches the expectation
        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), expected_response)

    def test_get_chat_data_blank_consumer(self):
        """Test case for blank consumer."""
        consumer_type = ''
        expected_interaction = {
            "errors": [
                {
                    "title": "BadRequest",
                    "detail": "Empty consumer id",
                    "status": "400"
                }
            ]
        }

        # Manually provide the token
        token = 'your_dynamic_token_here'

        # Define interaction for blank consumer
        self.define_interaction(
            description='A request with blank consumer',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            status_code=400,
            token=token
        )

        # Run the test: Make the actual request
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}', headers={"Authorization": f"Bearer {token}"})

        # Assert the 400 response for blank consumer
        self.assertEqual(result.status_code, 400)
        self.assertEqual(result.json(), expected_interaction)

    def test_get_chat_data_invalid_consumer(self):
        """Test case for an invalid consumer should return 200 with an empty template list."""
        consumer_type = 'invalid'
        expected_interaction = {
            "data": {
                "type": "ChatTemplates",
                "attributes": {
                    "templates": []
                }
            }
        }

        # Manually provide the token
        token = 'your_dynamic_token_here'

        # Define interaction for invalid consumer
        self.define_interaction(
            description='A request to get chat data for an invalid consumer',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            token=token
        )

        # Run the test: Make the actual request
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}', headers={"Authorization": f"Bearer {token}"})

        # Assert that the result matches the expectation
        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), expected_interaction)

    def test_get_chat_data_no_authorization(self):
        """Test case for missing Authorization header should return 401."""
        consumer_type = 'fullserve'
        expected_interaction = {
            "message": "Unauthorized"
        }

        # Define interaction for missing Authorization
        self.define_interaction(
            description='A request with no Authorization header',
            consumer_type=consumer_type,
            expected_response=expected_interaction,
            status_code=401,
            auth_required=False
        )

        # Run the test: Make the actual request without Authorization header
        with self.pact:
            result = requests.get(f'{pact.uri}/bcd-dev/actions/bcd/chat-template?consumer={consumer_type}')

        # Assert the 401 response for missing Authorization
        self.assertEqual(result.status_code, 401)
        self.assertEqual(result.json(), expected_interaction)


if __name__ == '__main__':
    unittest.main()
